### PR TITLE
fix: dont nuke trade input state when backing out of a trade preview

### LINKED ISSUE #6109: Don't fully reset trade input state when changing route
### Overview

Currently when clicking "back" from the trade confirmation the trade input state is totally cleared, preventing users from previewing a trade and combing back to it without re-entering everything.

Stale quotes should not be possible, so attention is needed to clear query state when coming back to trade input.

### References and additional details

In `src/components/MultiHopTrade/components/TradeInput/TradeInput.tsx`:
image

### Acceptance Criteria

Users can continue to use their trade input after previewing a quote and clicking back.
Stale quotes should not be possible.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Prevents clearing trade input state when previewing a trade and clicking "back".

Includes changes to manage input sell amounts in redux.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #6109

Moderate risk as this changes the behavior of trade input state between routes for trade input/preview/execution.

> What protocols, transaction types or contract interactions might be affected by this PR?

All trades.

## Testing

Check trade input behaves as expected
Ensure clicking "back" when previewing a trade maintains the previous input state
Ensure there is no way of somehow executing a trade with stale quotes or incorrect input data

### CODE REVIEW

**Comment by woodenfurniture** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/shapeshift/web/6113?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `develop`
  * **PR #6051** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6051?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #6077** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6077?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #6080** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6080?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #6113** <a href="https://app.graphite.dev/github/pr/shapeshift/web/6113?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/6113?utm_source=stack-comment).

File: src/components/MultiHopTrade/MultiHopTrade.tsx [hunk 1]
```diff
@@ -105,10 +89,11 @@ const MultiHopRoutes = memo(() => {
 
   useEffect(() => {
     return () => {
-      // Reset the swapper slice to initial state on mount
+      // Reset the trade slices to initial state on unmount
       // Don't move me to one of the trade route components, this needs to be at router-level
       // We only want to clear swapper state when trade components are fully unmounted, not when trade routes change
       dispatch(tradeInput.actions.clear())
+      dispatch(tradeQuoteSlice.actions.clear())
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/MultiHopTrade.tsx:96]: added this for safety/paranoia, though doesnt seem to be strictly required

File: src/components/MultiHopTrade/components/MultiHopTradeConfirm/MultiHopTradeConfirm.tsx [hunk 2]
```diff
@@ -33,7 +32,6 @@ export const MultiHopTradeConfirm = memo(() => {
   }, [dispatch, isLoading])
 
   const handleBack = useCallback(() => {
-    dispatch(swappersSlice.actions.clear())
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/components/MultiHopTradeConfirm/MultiHopTradeConfirm.tsx:36]: first main change

File: src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx [hunk 3]
```diff
@@ -29,8 +33,9 @@ export const SellAssetInput = memo(
     percentOptions,
     ...rest
   }: SellAssetInputProps) => {
-    const [sellAmountUserCurrencyHuman, setSellAmountUserCurrencyHuman] = useState('0')
-    const [sellAmountCryptoPrecision, setSellAmountCryptoPrecision] = useState('0')
+    const sellAmountCryptoPrecision = useAppSelector(selectInputSellAmountCryptoPrecision)
+    const sellAmountUserCurrency = useAppSelector(selectInputSellAmountUserCurrency)
+
```
Comment (woodenfurniture) [inline src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx:38]: all the changes to this file are basically moving state management into redux

File: src/state/slices/tradeInputSlice/tradeInputSlice.ts [hunk 4]
```diff
@@ -53,6 +53,9 @@ export const tradeInput = createSlice({
       const isSameAsBuyAsset = asset.assetId === state.buyAsset.assetId
       if (isSameAsBuyAsset) state.buyAsset = state.sellAsset
 
+      // clear the sell amount
+      state.sellAmountCryptoPrecision = '0'
+
```
Comment (woodenfurniture) [inline src/state/slices/tradeInputSlice/tradeInputSlice.ts:58]: deletions to `src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx` are essentially regained with this line

**Review by 0xApotheosis — CHANGES_REQUESTED** [review]
Found a cheeky regression when switching between assets:

https://github.com/shapeshift/web/assets/97164662/4d6340d3-b71d-4d42-bb6d-59375f185919

File: src/components/MultiHopTrade/MultiHopTrade.tsx [hunk 5]
```diff
@@ -49,54 +48,39 @@ const MultiHopEntries = [
...
+  const defaultBuyAsset = useAppSelector(state => selectAssetById(state, defaultBuyAssetId ?? ''))
 
-    useEffect(() => {
-      dispatch(tradeInput.actions.clear())
-      if (defaultSellAsset) dispatch(tradeInput.actions.setSellAsset(defaultSellAsset))
-      if (routeBuyAsset) dispatch(tradeInput.actions.setBuyAsset(routeBuyAsset))
-      else if (defaultBuyAsset) dispatch(tradeInput.actions.setBuyAsset(defaultBuyAsset))
-    }, [defaultBuyAsset, defaultSellAsset, dispatch, routeBuyAsset])
+  useEffect(() => {
+    dispatch(tradeInput.actions.clear())
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/MultiHopTrade.tsx:67]: Is the removed logic for the sell asset setting on purpose here?

File: src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx [hunk 6]
```diff
@@ -40,10 +45,7 @@ export const SellAssetInput = memo(
     // this is separated from handleSellAssetInputChange to prevent resetting the input value to 0
     // when the market data changes between keystrokes
     const handleSellAssetInputChangeInner = useCallback(
-      (sellAmountUserCurrencyHuman: string, sellAmountCryptoPrecision: string) => {
-        setSellAmountUserCurrencyHuman(sellAmountUserCurrencyHuman)
-        setSellAmountCryptoPrecision(sellAmountCryptoPrecision)
-
+      (sellAmountCryptoPrecision: string) => {
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx:48]: Will need to test this ensure this doesn't break fiat input / fiat from crypto prorating

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
Tested locally:

### Simple backing out of trade preview on the default ETH-FOX pair (i.e without any fancy change from crypto to fiat and back, swap/change of assets etc) ✅ 


https://github.com/shapeshift/web/assets/17035424/9ec5f459-c5bd-400a-9949-a97edb001b6b

### Fiat prorating

Fiat selection is lost when backing out, but otherwise looks good

https://github.com/shapeshift/web/assets/17035424/730791f0-f73a-4c9b-9844-f163d72c0c8d

### Assets switch

https://github.com/shapeshift/web/assets/17035424/ce2a0f41-43fb-4350-9b95-c4210ec6a7f4

As noted by @0xApotheosis, we lost reactivity here on assets *swap*, as we keep the amount of the previous sell asset. Assets *switch* however seems happy

### Route buy asset

image
image

- Note this is happy on refresh as well

**Review by gomesalexandre — COMMENTED** [review]
Retested locally after restack, noted issue WRT asset swaps while inputting fiat values still present

File: src/components/MultiHopTrade/MultiHopTrade.tsx (hunk 5 above)
Comment (woodenfurniture) [reply src/components/MultiHopTrade/MultiHopTrade.tsx:67]: yes it is, the store has a default asset for both buy and sell and only the default buy asset is overridden in this case

File: src/state/slices/tradeInputSlice/selectors.ts [hunk 7]
```diff
@@ -152,6 +152,15 @@ export const selectInputSellAmountUsd = createSelector(
   },
 )
 
+export const selectInputSellAmountUserCurrency = createSelector(
+  selectInputSellAmountCryptoPrecision,
+  selectInputSellAssetUserCurrencyRate,
+  (sellAmountCryptoPrecision, sellAssetUserCurrencyRate) => {
+    if (!sellAssetUserCurrencyRate) return
+    return bn(sellAmountCryptoPrecision).times(sellAssetUserCurrencyRate).toFixed()
```
[the remark this answers is not in the document]
Comment (woodenfurniture) [reply src/state/slices/tradeInputSlice/selectors.ts:174]: yeah input sell amount should be getting converted back into crypto precision before getting set in the slice, so provided this is the case it should be correct

**Review by 0xApotheosis — APPROVED** [review]
Feedback address and bugs squashed 👌

**Review by gomesalexandre — APPROVED** [review]
Retested locally:

- Confirmed the previously noted issue WRT fiat prorate is fixed and we zero-out inputs the same as in develop
- The previously tested flow of backing out of trade confirm i.e the guts of this PR, is still happy

File: src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx (hunk 6 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/TradeInput/components/SellAssetInput.tsx:50]: Confirmed happy 🤌

**Pull request merged**